[SPARK-26969][SQL] Using ODBC client not able to see the query data when column datatype is decimal

sujith71955 opened the pull request:
## What changes were proposed in this pull request?
While processing the Rowdata in the server side ColumnValue BigDecimal type value processed by server has to converted to the HiveDecmal data type for successful processing of query using Hive ODBC client.As per current logic  corresponding to the Decimal column datatype, the Spark server uses BigDecimal, and the ODBC client uses HiveDecimal. If the data type does not match, the client fail to parse

Since this handing was missing the query executed in Hive ODBC client wont return or provides result to the user even though the decimal type column value data present.

## How was this patch tested?

Manual test report and impact assessment is done using existing test-cases

Before fix 
![decimal_odbc](https://user-images.githubusercontent.com/12999161/53440179-e74a7f00-3a29-11e9-93db-83f2ae37ef16.PNG)

After Fix
![hive_odbc](https://user-images.githubusercontent.com/12999161/53679519-70e0a200-3cf3-11e9-9437-9c27d2e5056d.PNG)

sujith71955 committed:
[SPARK-26969][TSpark-Thriftserver] Bigdecimal values shall be converted to HiveDecimal type for succesful processing of query using ODBC client

## What changes were proposed in this pull request?
While processing the Rowdata in the server side ColumnValue BigDecimal type value processed by server has to converted to the HiveDecmal data type for successful processing of query using ODBC client.
Since this handing was missing the query executed in ODBC client wont return or provides result to the user even though the decimal type column value data present.

## How was this patch tested?

Manual test report and impact assessment is done using existing test-cases

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SongYadong commented on sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/ColumnValue.java:
```diff
@@ -207,6 +207,14 @@ public static TColumnValue toTColumnValue(Type type, Object value) {
+  private static TColumnValue getDecimalType(Object value) {
+    Object decV = value;
+      if (value instanceof BigDecimal) {
+        decV = HiveDecimal.create((BigDecimal)decV);
```
nit: the indent can be improved :)

sujith71955 replied:
fixed. thanks :)

sujith71955 commented:
cc @davies @HyukjinKwon

wangyum commented:
@sujith71955 Thanks for this fix. Could you describe how to reproduce this issue.  Which 
ODBC client? which ODBC driver? I need verify it first.

sujith71955 commented:
> @sujith71955 Thanks for this fix. Could you describe how to reproduce this issue. Which
> ODBC client? which ODBC driver? I need verify it first.

Its Hive ODBC client and UnixODBC driver. Further details you can refer the JIRA steps. In JIRA the steps has already added by thhe tester. Moreover you can also check in data-bricks site to know more about HiveODBC conenction details. Please revert me for any clarifications further. Thanks for your input ;)

sujith71955 commented:
@dongjoon-hyun @davies @HyukjinKwon @cloud-fan  Can i get any inputs regarding my changes for this JIRA. Thanks.

sujith71955 commented:
@wangyum are you able to verify? let me know if any clarifications you have.

sujith71955 commented:
cc @srowen

wangyum commented:
@sujith71955 I'll verify it next few days.

sujith71955 commented:
> @sujith71955 I'll verify it next few days.

Thanks wangyum

sujith71955 commented:
@wangyum @HyukjinKwon  Can i get some inputs for this PR. Thanks in advance

sujith71955 commented:
retest this please

cloud-fan commented:
add to whitelist

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #103322 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103322/testReport)** for PR 23899 at commit [`90109a3`](https://github.com/apache/spark/commit/90109a31fd14c2bd3aacb7dd2af03e5d4d6eb2ee).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8761/
Test PASSed.

SparkQA commented:
**[Test build #103322 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103322/testReport)** for PR 23899 at commit [`90109a3`](https://github.com/apache/spark/commit/90109a31fd14c2bd3aacb7dd2af03e5d4d6eb2ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103322/
Test PASSed.

sujith71955 commented:
@cloud-fan @dongjoon-hyun - please review and let me know for any suggestions. Thanks

sujith71955 commented:
@cloud-fan @dongjoon-hyun - please let me know for  any suggestion on the above PR as directly the querues with decimal type column are not giving any results in hive client. thanks

sujith71955 commented:
gentle ping @cloud-fan @dongjoon-hyun

sujith71955 commented:
gentle ping @cloud-fan @dongjoon-hyun

sujith71955 commented:
gentle ping @cloud-fan @dongjoon-hyun

wangyum commented:
@sujith71955 Could you please try these ODBC drivers:
https://docs.hortonworks.com/HDPDocuments/HDP2/HDP-2.6.5/bk_data-access/content/hive-jdbc-odbc-drivers.html
https://www.cloudera.com/documentation/other/connectors/hive-odbc/latest.html

sujith71955 commented:
@cloud-fan @dongjoon-hyun Kindly review and let me know for your opinion.
@wangyum I will let you know the results. Did you verified? This happens only in case of ODBC driver, because HIVE_CLI_SERVICE_PROTOCOL_V1 so it goes with RowBasedSet.

sujith71955 commented:
gentle ping @cloud-fan @dongjoon-hyun  - please review this PR and let me know for any clarification. the usecase is directly failing if user try to execute a decimal type column using ODBC client. thanks

sujith71955 commented:
gentle ping @cloud-fan @dongjoon-hyun

sujith71955 commented:
gentle ping @cloud-fan @dongjoon-hyun

sujith71955 commented:
gentle ping @cloud-fan @dongjoon-hyun

sujith71955 commented:
@dongjoon-hyun @cloud-fan @HyukjinKwon  Gentle reminder for the review. thanks

sujith71955 commented:
@dongjoon-hyun @cloud-fan Gentle reminder for the review. thanks

dongjoon-hyun commented on sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/ColumnValue.java:
```diff
@@ -193,7 +193,7 @@ public static TColumnValue toTColumnValue(Type type, Object value) {
     case INTERVAL_DAY_TIME_TYPE:
       return stringValue((HiveIntervalDayTime) value);
     case DECIMAL_TYPE:
-      return stringValue(((HiveDecimal)value));
```
Sorry for late response, @sujith71955 . But, this PR is difficult to validate.

BTW, Hive seems to choose a different path; [here](https://github.com/apache/hive/blob/master/service/src/java/org/apache/hive/service/cli/ColumnValue.java#L210-L211) and [here](https://github.com/apache/hive/blob/master/service/src/java/org/apache/hive/service/cli/ColumnValue.java#L151-L158)
```java
    case DECIMAL_TYPE:
      return stringValue((HiveDecimal)value, typeDescriptor);
```

```java
  private static TColumnValue stringValue(HiveDecimal value, TypeDescriptor typeDescriptor) {
    TStringValue tStrValue = new TStringValue();
    if (value != null) {
      int scale = typeDescriptor.getDecimalDigits();
      tStrValue.setValue(value.toFormatString(scale));
    }
    return TColumnValue.stringVal(tStrValue);
  }
```

Since this PR reports a bug with `Hive ODBC Client`, I'm wondering if Hive works correctly with Hive ODBC Driver with the above code? If then, isn't it more natural to follow the upstream?

sujith71955 replied:
i will explore more based on your comment and update you guys.Thanks for your valuable suggestion.

dongjoon-hyun commented:
Gentle ping, @sujith71955 .

wangyum commented on sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/ColumnValue.java:
```diff
@@ -193,7 +193,7 @@ public static TColumnValue toTColumnValue(Type type, Object value) {
       return stringValue((HiveIntervalDayTime) value);
     case DECIMAL_TYPE:
-      return stringValue(((HiveDecimal)value));
+      return getDecimalType(value);
```
The reason is your client protocol version is lower than `HIVE_CLI_SERVICE_PROTOCOL_V6`:
https://github.com/apache/spark/blob/02c33694c8254f69cb36c71c0876194dccdbc014/sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/RowSetFactory.java#L28-L40

The fix is correct, but I'd like make it more clear:
```diff
+++ b/sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/ColumnValue.java
@@ -23,7 +23,6 @@ import java.sql.Date;
 import java.sql.Timestamp;
 
 import org.apache.hadoop.hive.common.type.HiveChar;
-import org.apache.hadoop.hive.common.type.HiveDecimal;
 import org.apache.hadoop.hive.common.type.HiveIntervalDayTime;
 import org.apache.hadoop.hive.common.type.HiveIntervalYearMonth;
 import org.apache.hadoop.hive.common.type.HiveVarchar;
@@ -139,11 +138,10 @@ public class ColumnValue {
     return TColumnValue.stringVal(tStringValue);
   }
 
-  private static TColumnValue stringValue(HiveDecimal value, TypeDescriptor typeDescriptor) {
+  private static TColumnValue stringValue(BigDecimal value, TypeDescriptor typeDescriptor) {
     TStringValue tStrValue = new TStringValue();
     if (value != null) {
-      int scale = typeDescriptor.getDecimalDigits();
-      tStrValue.setValue(value.toFormatString(scale));
+      tStrValue.setValue(value.toPlainString());
     }
     return TColumnValue.stringVal(tStrValue);
   }
@@ -197,7 +195,7 @@ public class ColumnValue {
     case INTERVAL_DAY_TIME_TYPE:
       return stringValue((HiveIntervalDayTime) value);
     case DECIMAL_TYPE:
-      return stringValue((HiveDecimal)value, typeDescriptor);
+      return stringValue((BigDecimal)value, typeDescriptor);
     case BINARY_TYPE:
       return stringValue((String)value);
     case ARRAY_TYPE:
```

We can verify it by https://github.com/apache/spark/pull/25228

wangyum commented:
@sujith71955 Cloud you update the PR and enable this test case:
https://github.com/apache/spark/blob/a59fdc4b5783be591a236bfc60d1107caa818412/sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala#L186-L193

SparkQA commented:
**[Test build #108808 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108808/testReport)** for PR 23899 at commit [`97c4ca5`](https://github.com/apache/spark/commit/97c4ca5c5d6d6c677737e8ffe4659c5de9b30dca).

sujith71955 commented:
> @sujith71955 Cloud you update the PR and enable this test case:
> https://github.com/apache/spark/blob/a59fdc4b5783be591a236bfc60d1107caa818412/sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala#L186-L193

Done. Thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13889/
Test PASSed.

sujith71955 committed:
[SPARK-26969][TSpark-Thriftserver] Bigdecimal values shall be converted to HiveDecimal type for succesful processing of query using ODBC client

## What changes were proposed in this pull request?
While processing the Rowdata in the server side ColumnValue BigDecimal type value processed by server has to converted to the HiveDecmal data type for successful processing of query using ODBC client.
Since this handing was missing the query executed in ODBC client wont return or provides result to the user even though the decimal type column value data present.

## How was this patch tested?

Manual test report and impact assessment is done using existing test-cases

SparkQA commented:
**[Test build #108809 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108809/testReport)** for PR 23899 at commit [`64e3031`](https://github.com/apache/spark/commit/64e303122607a2b0a7beeb03a2c95b820d36b1d5).

wangyum commented on sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/ColumnValue.java:
```diff
@@ -195,7 +195,7 @@ public static TColumnValue toTColumnValue(Type type, Object value) {
       return stringValue((HiveIntervalDayTime) value);
     case DECIMAL_TYPE:
-      return stringValue(((HiveDecimal)value));
+      return getDecimalType(value);
```
```return stringValue(((BigDecimal)value).toPlainString());```?


https://github.com/apache/hive/blob/release-2.3.5-rc0/storage-api/src/java/org/apache/hadoop/hive/common/type/HiveDecimal.java#L725

wangyum replied:
We also need to update v2.3.5's `ColumnValue`: https://github.com/apache/spark/blob/master/sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/ColumnValue.java#L201-L202

sujith71955 replied:
> `return stringValue(((BigDecimal)value).toPlainString());`?
> 
> https://github.com/apache/hive/blob/release-2.3.5-rc0/storage-api/src/java/org/apache/hadoop/hive/common/type/HiveDecimal.java#L725

you mean getDecimalType() function is not required?

wangyum replied:
Yes.

sujith71955 replied:
k will update  just need one clarification 'toPlainString()' conversion takes care necessariy normalizations as all? because create function does that

![image](https://user-images.githubusercontent.com/12999161/62694050-38a36500-b9f1-11e9-9853-9b1b755c755d.png)

sujith71955 replied:
ultimately it converts to HiveDecimal and then we are sending back the string value.

wangyum replied:
Please avoid using `HiveDecimal` because it may cause inconsistent results for `ColumnBasedSet` implementations and `RowBasedSet` implementations:
https://github.com/apache/spark/blob/02c33694c8254f69cb36c71c0876194dccdbc014/sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/RowSetFactory.java#L35-L40

sujith71955 replied:
i didn't got exactly pardon me , but if client expects HiveDecimal then?

SparkQA commented:
**[Test build #108808 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108808/testReport)** for PR 23899 at commit [`97c4ca5`](https://github.com/apache/spark/commit/97c4ca5c5d6d6c677737e8ffe4659c5de9b30dca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108808/
Test PASSed.

SparkQA commented:
**[Test build #108809 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108809/testReport)** for PR 23899 at commit [`64e3031`](https://github.com/apache/spark/commit/64e303122607a2b0a7beeb03a2c95b820d36b1d5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108809/
Test FAILed.

sujith71955 committed:
[SPARK-26969][TSpark-Thriftserver] Bigdecimal values shall be converted to HiveDecimal type for succesful processing of query using ODBC client

## What changes were proposed in this pull request?
While processing the Rowdata in the server side ColumnValue BigDecimal type value processed by server has to converted to the HiveDecmal data type for successful processing of query using ODBC client.
Since this handing was missing the query executed in ODBC client wont return or provides result to the user even though the decimal type column value data present.

## How was this patch tested?

Manual test report and impact assessment is done using existing test-cases

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13900/
Test PASSed.

SparkQA commented:
**[Test build #108822 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108822/testReport)** for PR 23899 at commit [`85931c8`](https://github.com/apache/spark/commit/85931c8e455eb0f76226bb6e079e36f4040d8d27).

wangyum commented on sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/ColumnValue.java:
```diff
@@ -195,7 +195,7 @@ public static TColumnValue toTColumnValue(Type type, Object value) {
+      return stringValue(((BigDecimal)value).toPlainString());
     case BINARY_TYPE:
       String strVal = value == null ? null : UTF8String.fromBytes((byte[])value).toString();
       return stringValue(strVal);
```
Please remove these lines because we do not need it:
https://github.com/apache/spark/pull/23899/files#diff-15f177f5340d8433fafc231afa298b53R26
https://github.com/apache/spark/pull/23899/files#diff-15f177f5340d8433fafc231afa298b53R143-R149

sujith71955 replied:
done

sujith71955 replied:
done

SparkQA commented:
**[Test build #108822 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108822/testReport)** for PR 23899 at commit [`85931c8`](https://github.com/apache/spark/commit/85931c8e455eb0f76226bb6e079e36f4040d8d27).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108822/
Test PASSed.

sujith71955 committed:
[SPARK-26969][TSpark-Thriftserver] Bigdecimal values shall be converted to HiveDecimal type for succesful processing of query using ODBC client

## What changes were proposed in this pull request?
While processing the Rowdata in the server side ColumnValue BigDecimal type value processed by server has to converted to the HiveDecmal data type for successful processing of query using ODBC client.
Since this handing was missing the query executed in ODBC client wont return or provides result to the user even though the decimal type column value data present.

## How was this patch tested?

Manual test report and impact assessment is done using existing test-cases

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13901/
Test PASSed.

SparkQA commented:
**[Test build #108824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108824/testReport)** for PR 23899 at commit [`5b4331e`](https://github.com/apache/spark/commit/5b4331eec6428cb5c834c8b634eaff973765fea9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13902/
Test PASSed.

wangyum commented on sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/ColumnValue.java:
```diff
@@ -199,7 +199,7 @@ public static TColumnValue toTColumnValue(TypeDescriptor typeDescriptor, Object
-      return stringValue((HiveDecimal)value, typeDescriptor);
+      stringValue(((BigDecimal)value).toPlainString());
     case BINARY_TYPE:
       String strVal = value == null ? null : UTF8String.fromBytes((byte[])value).toString();
```
Could you remove these lines:
https://github.com/apache/spark/pull/23899/files#diff-5aed8ed671b217c7920c373fc491869aL26
https://github.com/apache/spark/pull/23899/files#diff-5aed8ed671b217c7920c373fc491869aR144-R151

SparkQA commented:
**[Test build #108824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108824/testReport)** for PR 23899 at commit [`5b4331e`](https://github.com/apache/spark/commit/5b4331eec6428cb5c834c8b634eaff973765fea9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108824/
Test PASSed.

sujith71955 committed:
[SPARK-26969][TSpark-Thriftserver] Bigdecimal values shall be converted to HiveDecimal type for succesful processing of query using ODBC client

## What changes were proposed in this pull request?
While processing the Rowdata in the server side ColumnValue BigDecimal type value processed by server has to converted to the HiveDecmal data type for successful processing of query using ODBC client.
Since this handing was missing the query executed in ODBC client wont return or provides result to the user even though the decimal type column value data present.

## How was this patch tested?

Manual test report and impact assessment is done using existing test-cases

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13903/
Test PASSed.

SparkQA commented:
**[Test build #108825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108825/testReport)** for PR 23899 at commit [`c701a71`](https://github.com/apache/spark/commit/c701a7165507be906cf3cf82bbee9aadcbdb197f).

SparkQA commented:
**[Test build #108825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108825/testReport)** for PR 23899 at commit [`c701a71`](https://github.com/apache/spark/commit/c701a7165507be906cf3cf82bbee9aadcbdb197f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108825/
Test PASSed.

wangyum commented:
Thank you @sujith71955 Please update PR title according to the PR content.
Also add `test-hadoop3.2` to PR title to test with our [v2.3.5](https://github.com/apache/spark/tree/master/sql/hive-thriftserver/v2.3.5).

sujith71955 commented:
> Thank you @sujith71955 Please update PR title according to the PR content.
> Also add `test-hadoop3.2` to PR title to test with our [v2.3.5](https://github.com/apache/spark/tree/master/sql/hive-thriftserver/v2.3.5).

Done.

sujith71955 commented:
Thanks for the inputs.

wangyum commented:
retest this please

SparkQA commented:
**[Test build #108863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108863/testReport)** for PR 23899 at commit [`c701a71`](https://github.com/apache/spark/commit/c701a7165507be906cf3cf82bbee9aadcbdb197f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13936/
Test PASSed.

SparkQA commented:
**[Test build #108863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108863/testReport)** for PR 23899 at commit [`c701a71`](https://github.com/apache/spark/commit/c701a7165507be906cf3cf82bbee9aadcbdb197f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108863/
Test PASSed.

wangyum commented:
LGTM
cc @HyukjinKwon @dongjoon-hyun

dongjoon-hyun reviewed (approved):
+1, LGTM, too. 
Thank you so much, @sujith71955 , @wangyum , @SongYadong , @cloud-fan .
Merged to master.

HyukjinKwon commented:
Thanks @dongjoon-hyun and thanks for cc'ing me @wangyum. LGTM too.

sujith71955 commented:
Thanks all for the guidance.
